test/test_euler.py

```diff
@@ -0,0 +1,874 @@
+from mirgecom.initializers import Vortex2D
+from mirgecom.initializers import Lump
+from mirgecom.boundary import PrescribedBoundary
+from mirgecom.eos import IdealSingleGas
+from mirgecom.integrators import rk4_step
+from grudge.eager import EagerDGDiscretization
+from pyopencl.tools import (  # noqa
+    pytest_generate_tests_for_pyopencl as pytest_generate_tests,
+)
+
+
+def test_inviscid_flux():
+    """Checks that the Euler-internal inviscid flux
+    routine (_inviscid_flux) returns exactly the
+    expected result. The test checks flux for 1,
+    2, and 3 spatial dimensions.
+
+    The test is in 3 parts:
+
+    - Exact expression: Checks the returned flux
+      against the exact expressions:
+      F(q) = <rhoV, (E+p)V, rho(V.x.V) + pI>
+    - Prescribed p, V = 0:  Checks that only
+      diagonal terms of the momentum flux
+      [ rho(V.x.V) + pI ] are non-zero and return
+      the correctly calculated p.
+    - Prescribed p, V != 0: Checks that the
+      flux terms are returned in the proper
+      order by running only 1 non-zero velocity
+      component at-a-time.
+    """
+    ctx = cl.create_some_context(interactive=False)
+    queue = cl.CommandQueue(ctx)
+
+    cl_ctx = cl.create_some_context()
+    queue = cl.CommandQueue(cl_ctx)
+
+    dim = 2
+    nel_1d = 16
+
+    from meshmode.mesh.generation import generate_regular_rect_mesh
+
+    for dim in [1, 2, 3]:
+        mesh = generate_regular_rect_mesh(
+            a=(-0.5,) * dim, b=(0.5,) * dim, n=(nel_1d,) * dim
+        )
+
+        order = 3
+        discr = EagerDGDiscretization(cl_ctx, mesh, order=order)
+        eos = IdealSingleGas()
+
+        logging.info(f"Number of {dim}d elems: {mesh.nelements}")
+
+        mass = cl.clrandom.rand(
+            queue, (mesh.nelements,), dtype=np.float64
+        )
+        energy = cl.clrandom.rand(
+            queue, (mesh.nelements,), dtype=np.float64
+        )
+        mom = make_obj_array(
+            [
+                cl.clrandom.rand(
+                    queue, (mesh.nelements,), dtype=np.float64
+                )
+                for i in range(dim)
+            ]
+        )
+
+        q = flat_obj_array(mass, energy, mom)
+
+        # Create the expected result
+        p = eos.pressure(q)
+        escale = (energy + p) / mass
+        expected_mass_flux = mom
+        expected_energy_flux = mom * make_obj_array([escale])
+        expected_mom_flux = make_obj_array(
+            [
+                (mom[i] * mom[j] / mass + (p if i == j else 0))
+                for i in range(dim)
+                for j in range(dim)
+            ]
+        )
+        expected_flux = flat_obj_array(
+            expected_mass_flux, expected_energy_flux, expected_mom_flux
+        )
+
+        from mirgecom.euler import _inviscid_flux
+
+        flux = _inviscid_flux(discr, q, eos)
+        flux_resid = flux - expected_flux
+
+        for i in range((dim + 2) * dim):
+            assert (la.norm(flux_resid[i].get())) == 0.0
```

Comment:
If I'm seeing this right, this kind of just seems to duplicate the formulas that are already present in the Euler submodule, so it's not that much of an independent verification to me. Could you explain the thought process (or tell me that I'm reading this wrong)?

Reply:
TDD naturally led me to this test as I laid down and tested field-by-field. It (i.e. the first test in this routine) seems like an odd test now that it is all done and passing in its final state but it serves as a sort of identity test that will fail if the flux routine is "refactored" away from the expected result.